### ISSUE TITLE
No HDMI output in firmware

### ISSUE THREAD

**Issue by johanes2115** [Dasharo/dasharo-issues#533]
### Device

Novacustom TGL

### Dasharo version

v1.7.0

### Affected component(s) or functionality

HDMI output

### Brief summary

There is no HDMI output in firmware, which makes remote automatic validation not possible.

### How reproducible

100%

### How to reproduce

1. Connect external HDMI output
2. Power on the device

### Expected behavior

HDMI output in firmware.

### Actual behavior

No HDMI output in firmware.

**Comment by miczyg1** [comment]
Unfixable. TGL FSP creates the GOPPolicy with GetPlatformLidStatus returning EFI_UNSUPPORTED. So unless you hack your own FSP and expose such option, this is not doable.

**Comment by pietrushnic** [comment]
@miczyg1 AFAIK, we should be able to do that. Another reason is to choose Dasharo.

**Comment by miczyg1** [comment]
> @miczyg1 AFAIK, we should be able to do that. Another reason is to choose Dasharo.

Yes, we can.

**Comment by mkopec** [comment]
@miczyg1 could we not just include the Intel GOP driver in UEFIPayload and run it with policy specified by UEFI instead of coreboot?

**Comment by mkopec** [comment]
> If we really need this, we can just unplug the eDP cable, FSP will see it's missing and output to HDMI 

Yeah, I needed KVM to debug issues with testing quicker, and just unplugged the eDP cable and I get FW on HDMI.

**Comment by macpijan** [comment]
> v1.7.0

The author meant 1.5.0 probably.

This is most likely still relevant, as it won't fix itself out of nowhere. No need to recheck it, IMHO. We can go back to actually implementing it, if we decide to do so.

**Comment by mkopec** [comment]
Can be fixed the same way we fixed it on meteorlake, by adding the GOP and policy drivers to edk2.

**Comment by BeataZdunczyk** [comment]
OK, we will schedule fixing this issue before the next release.

**Comment by macpijan** [comment]
> OK, we will schedule fixing this issue before the next release.

Let's confirm with @wessel-novacustom on the priority

**Comment by miczyg1** [comment]
> @miczyg1 could we not just include the Intel GOP driver in UEFIPayload and run it with policy specified by UEFI instead of coreboot?

Most likely.

**Comment by JakubRed** [comment]
This issue was tested on NV40MB, NV41MB, NS70MU, NS50MU, NV41PZ, NS51PU, NS70PU, V560TU and V540TU on the most recent public firmware. Every time there was a reproduction (No HDMI output in firmware). Which indicates that the issue still persists on all platforms.

**Comment by miczyg1** [comment]
> This issue was tested on NV40MB, NV41MB, NS70MU, NS50MU, NV41PZ, NS51PU, NS70PU, V560TU and V540TU on the most recent public firmware. Every time there was a reproduction (No HDMI output in firmware). Which indicates that the issue still persists on all platforms.

So basically all models? Rather unlikely. The  NV41PZ, NS51PU, NS70PU, V560TU and V540TU at least should have a working HDMI in firmware. The other ones do not support it AFAIK. Just to be clear. Did you boot the laptops with lid closed?

**Comment by mkopec** [comment]
@filipleple also confirmed it doesn't work on the latest code

**Comment by miczyg1** [comment]
> [@filipleple](https://github.com/filipleple) also confirmed it doesn't work on ADL on the latest code,so something must've broken. It does work on our work laptops after all

IIRC we've switched RPL FSP, so that may be the cause?

I saw on matrix that DMA protection also doesn't work on MSI on latest release. I wouldn't be surprised if Intel dropped a couple of things in RPL FSP that were present in ADL FSP.

**Comment by PerAstraAdDeum** [comment]
> > This issue was tested on NV40MB, NV41MB, NS70MU, NS50MU, NV41PZ, NS51PU, NS70PU, V560TU and V540TU on the most recent public firmware. Every time there was a reproduction (No HDMI output in firmware). Which indicates that the issue still persists on all platforms.
> 
> So basically all models? Rather unlikely. The NV41PZ, NS51PU, NS70PU, V560TU and V540TU at least should have a working HDMI in firmware. The other ones do not support it AFAIK. Just to be clear. Did you boot the laptops with lid closed?

I can confirm that there is no HDMI output when in UEFI menu on the V540TU (Intel-only), if that is what this is about.

**Comment by miczyg1** [comment]
> > So basically all models? Rather unlikely. The NV41PZ, NS51PU, NS70PU, V560TU and V540TU at least should have a working HDMI in firmware. The other ones do not support it AFAIK. Just to be clear. Did you boot the laptops with lid closed?
> 
> No. I connected the HDMI before powering on but didn't boot with lid closed.

The firmware will output on HDMI only when the LID is closed. Otherwise firmware will initialize the display only on the LCD panel...

**Comment by miczyg1** [comment]
> > > This issue was tested on NV40MB, NV41MB, NS70MU, NS50MU, NV41PZ, NS51PU, NS70PU, V560TU and V540TU on the most recent public firmware. Every time there was a reproduction (No HDMI output in firmware). Which indicates that the issue still persists on all platforms.
> > 
> > 
> > So basically all models? Rather unlikely. The NV41PZ, NS51PU, NS70PU, V560TU and V540TU at least should have a working HDMI in firmware. The other ones do not support it AFAIK. Just to be clear. Did you boot the laptops with lid closed?
> 
> I can confirm that there is no HDMI output when in UEFI menu on the V540TU (Intel-only), if that is what this is about.

Yes, this is correct. On V540TU it is tricky to do it, because the power button is near the keyboard, unlike on previous models on the side of the laptop. So one really can't power on the laptop with lid closed. But, as a workaround you may try
`sudo systemctl reboot --firmware` in Linux and close the lid immediately after pressing enter. That should make the firmware enter UEFI setup automatically and hopefully display on HDMI will show up.

**Comment by miczyg1** [comment]
I guess we need to add that information to test procedure: https://docs.dasharo.com/unified-test-documentation/dasharo-compatibility/31E-display-ports-and-lcd/

Since I have added the basic display test procedures for Protectli (https://github.com/Dasharo/docs/commit/aa54ee455b2935abdf6dffb3c12ffa2b63547446), nobody revisisted it later, e.g. for laptops

**Comment by mkopec** [comment]
Fixed by https://github.com/Dasharo/coreboot/pull/617
